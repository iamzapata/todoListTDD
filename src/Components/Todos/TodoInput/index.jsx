import React, { PureComponent } from "react"
import { func } from "prop-types"

class TodoInput extends PureComponent {
  state = {
    inputText: ""
  }

  onChangeInput = ev => {
    const { value: inputText } = ev.target
    this.setState({ inputText })
  }

  addTodo = ev => {
    ev.preventDefault()
    const { inputText } = this.state
    const { addTodo } = this.props
    if (!inputText) return
    addTodo(inputText)
    this.setState({ inputText: "" })
  }

  render() {
    const { inputText } = this.state
    return (
      <form onSubmit={this.addTodo} className="TodoInput">
        <input
          type="text"
          placeholder="Add Todo..."
          value={inputText}
          onChange={this.onChangeInput}
        />
      </form>
    )
  }
}

TodoInput.propTypes = {
  addTodo: func.isRequired
}

export default TodoInput
