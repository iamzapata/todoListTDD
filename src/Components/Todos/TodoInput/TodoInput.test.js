import React from "react"
import { render, cleanup, fireEvent } from "react-testing-library"
import TodoInput from "./index"

describe("<TodoInput>", () => {
  afterEach(cleanup)

  const createProps = () => ({
    addTodo: jest.fn()
  })

  const todoPlaceholder = "Add Todo..."

  it("Should render self", () => {
    const { container } = render(<TodoInput {...createProps()} />)

    expect(container.firstChild).toBeTruthy()
    expect(container.firstChild).toMatchSnapshot()
  })

  it("Should have a TodoInput className", () => {
    const { container } = render(<TodoInput {...createProps()} />)
    const classNameSelection = container.querySelectorAll(".TodoInput")

    expect(classNameSelection.length).toEqual(1)
  })

  it("Should have a form element", () => {
    const { container } = render(<TodoInput {...createProps()} />)
    const form = container.querySelector("form")

    expect(form).toBeTruthy()
  })

  it("Should have an input element", () => {
    const { container } = render(<TodoInput {...createProps()} />)
    const input = container.querySelector("input")

    expect(input).toBeTruthy()
  })

  it("Should allow a user to type a task to add", () => {
    const task = "Research machine learning"
    const { queryByPlaceholderText } = render(<TodoInput {...createProps()} />)
    const inputNode = queryByPlaceholderText(todoPlaceholder)

    inputNode.value = task
    fireEvent.change(inputNode)

    expect(queryByPlaceholderText(todoPlaceholder).value).toEqual(task)
  })

  it("Should allow the user to create todo by pressing the enter key", () => {
    const props = createProps()
    const { queryByPlaceholderText } = render(<TodoInput {...props} />)
    const inputNode = queryByPlaceholderText(todoPlaceholder)

    inputNode.value = "Practice playing guitar"
    fireEvent.change(inputNode)
    fireEvent.submit(inputNode)

    expect(props.addTodo).toHaveBeenCalledTimes(1)
  })

  it("Should clear input when todo has been submitted", () => {
    const { queryByPlaceholderText } = render(<TodoInput {...createProps()} />)
    const inputNode = queryByPlaceholderText(todoPlaceholder)

    inputNode.value = "Do some cooking"
    fireEvent.change(inputNode)
    fireEvent.submit(inputNode)

    expect(queryByPlaceholderText(todoPlaceholder).value).toBeFalsy()
  })

  it("Should not allow to submit an empty todo", () => {
    const props = createProps()
    const { queryByPlaceholderText } = render(<TodoInput {...props} />)
    const inputNode = queryByPlaceholderText(todoPlaceholder)

    inputNode.value = ""
    fireEvent.change(inputNode)
    fireEvent.submit(inputNode)

    expect(props.addTodo).not.toBeCalled()
  })
})
