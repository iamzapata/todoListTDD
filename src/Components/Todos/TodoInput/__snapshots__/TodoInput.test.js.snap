// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoInput> Should render self 1`] = `
<form
  class="TodoInput"
>
  <input
    placeholder="Add Todo..."
    type="text"
    value=""
  />
</form>
`;
